// @datorama/akita/esm2015/src/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const currentAction = {
    type: null,
    entityIds: null,
    skip: false
};
/** @type {?} */
let customActionActive = false;
/**
 * @return {?}
 */
export function resetCustomAction() {
    customActionActive = false;
}
// public API for custom actions. Custom action always wins
/**
 * @param {?} type
 * @param {?=} entityIds
 * @return {?}
 */
export function logAction(type, entityIds) {
    setAction(type, entityIds);
    customActionActive = true;
}
/**
 * @param {?} type
 * @param {?=} entityIds
 * @return {?}
 */
export function setAction(type, entityIds) {
    if (customActionActive === false) {
        currentAction.type = type;
        currentAction.entityIds = entityIds;
    }
}
/**
 * @param {?=} skip
 * @return {?}
 */
export function setSkipAction(skip = true) {
    currentAction.skip = skip;
}
/**
 * @param {?} action
 * @param {?=} entityIds
 * @return {?}
 */
export function action(action, entityIds) {
    return (/**
     * @param {?} target
     * @param {?} propertyKey
     * @param {?} descriptor
     * @return {?}
     */
    function (target, propertyKey, descriptor) {
        /** @type {?} */
        const originalMethod = descriptor.value;
        descriptor.value = (/**
         * @param {...?} args
         * @return {?}
         */
        function (...args) {
            logAction(action, entityIds);
            return originalMethod.apply(this, args);
        });
        return descriptor;
    });
}
