// @datorama/akita/esm2015/src/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export {} from './types';
export { EntityStore, EntityUIStore } from './entityStore';
export { QueryEntity, EntityUIQuery } from './queryEntity';
export { Query } from './query';
export { Store } from './store';
export { applyTransaction, transaction, commit, endBatch, isTransactionInProcess, startBatch, transactionManager, withTransaction } from './transaction';
export { filterNil } from './filterNil';
export { DEFAULT_ID_KEY } from './defaultIDKey';
export { action, setAction, setSkipAction, logAction, currentAction, resetCustomAction } from './actions';
export { SnapshotManager, snapshotManager } from './snapshotManager';
export { configKey, StoreConfig } from './storeConfig';
export { QueryConfig, queryConfigKey } from './queryConfig';
export { akitaConfig, getAkitaConfig } from './config';
export { compareValues, Order } from './sort';
export { AkitaPlugin } from './plugins/plugin';
export { Paginator, PaginatorPlugin } from './plugins/paginator/paginatorPlugin';
export { PersistNgFormPlugin } from './plugins/persistForm/persistNgFormPlugin';
export { persistState, selectPersistStateInit } from './persistState';
export { akitaDevtools } from './devtools';
export { EntityCollectionPlugin } from './plugins/entityCollectionPlugin';
export { StateHistoryPlugin } from './plugins/stateHistory/stateHistoryPlugin';
export { EntityStateHistoryPlugin } from './plugins/stateHistory/entityStateHistoryPlugin';
export { dirtyCheckDefaultParams, DirtyCheckPlugin, getNestedPath } from './plugins/dirtyCheck/dirtyCheckPlugin';
export { EntityDirtyCheckPlugin } from './plugins/dirtyCheck/entityDirtyCheckPlugin';
export { guid } from './guid';
export { setEntities, isEntityState } from './setEntities';
export { toEntitiesIds } from './toEntitiesIds';
export { toEntitiesObject } from './toEntitiesObject';
export { hasEntity } from './hasEntity';
export { hasActiveState, getExitingActives, isMultiActiveState, resolveActiveEntity } from './activeState';
export { isEmpty } from './isEmpty';
export { coerceArray } from './coerceArray';
export { updateEntities } from './updateEntities';
export { removeEntities, removeAllEntities } from './removeEntities';
export { getInitialEntitiesState } from './getInitialEntitiesState';
export { getActiveEntities } from './getActiveEntities';
export { addEntities } from './addEntities';
export { resetStores } from './resetStores';
export { isObject } from './isObject';
export { isPlainObject } from './isPlainObject';
export { isFunction } from './isFunction';
export { isArray } from './isArray';
export { toBoolean } from './toBoolean';
export { isUndefined } from './isUndefined';
export { isNil } from './isNil';
export { isString } from './isString';
export { isNumber } from './isNumber';
export { isDefined } from './isDefined';
export { setValue } from './setValueByString';
export { getValue } from './getValueByString';
export { sortByOptions } from './sortByOptions';
export { entitiesToArray } from './entitiesToArray';
export { entitiesToMap } from './entitiesToMap';
export { __stores__ } from './stores';
export { isDev, enableAkitaProdMode, __DEV__ } from './env';
export { isNotBrowser } from './root';
export { runStoreAction, StoreActions } from './runStoreAction';
export { arrayUpdate } from './arrayUpdate';
export { arrayAdd } from './arrayAdd';
export { arrayUpsert } from './arrayUpsert';
export { arrayFind, distinctUntilArrayItemChanged, find } from './arrayFind';
export { arrayRemove } from './arrayRemove';
export { createEntityQuery, createEntityStore, createQuery, createStore } from './fp';
export { EntityActions } from './entityActions';
export { dispatchDeleted, dispatchAdded, dispatchUpdate, $$deleteStore, $$addStore, $$updateStore } from './dispatchers';
export { cacheable } from './cacheable';
export { combineQueries } from './combineQueries';
export { EntityService } from './entityService';
export { setLoading } from './setLoading';
