// @datorama/akita-ngdevtools/esm5/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { APP_INITIALIZER, NgModule, NgZone, InjectionToken, Inject, Injectable } from '@angular/core';
import { akitaDevtools } from '@datorama/akita';
/** @type {?} */
export var DEVTOOLS_OPTIONS = new InjectionToken('DevtoolsOptions');
var AkitaDevtools = /** @class */ (function () {
    function AkitaDevtools(ngZone, options) {
        this.ngZone = ngZone;
        this.options = options;
        akitaDevtools(this.ngZone, this.options);
    }
    AkitaDevtools.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AkitaDevtools.ctorParameters = function () { return [
        { type: NgZone },
        { type: undefined, decorators: [{ type: Inject, args: [DEVTOOLS_OPTIONS,] }] }
    ]; };
    return AkitaDevtools;
}());
export { AkitaDevtools };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AkitaDevtools.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    AkitaDevtools.prototype.options;
}
// auto initialize the devtools
/**
 * @return {?}
 */
export function d() { }
/**
 * @param {?} akitaDevtools
 * @return {?}
 */
export function init(akitaDevtools) {
    return d;
}
var AkitaNgDevtools = /** @class */ (function () {
    function AkitaNgDevtools() {
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    AkitaNgDevtools.forRoot = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: AkitaNgDevtools,
            providers: [
                AkitaDevtools,
                {
                    provide: DEVTOOLS_OPTIONS,
                    useValue: options
                },
                {
                    provide: APP_INITIALIZER,
                    useFactory: init,
                    deps: [AkitaDevtools],
                    multi: true
                }
            ]
        };
    };
    AkitaNgDevtools.decorators = [
        { type: NgModule, args: [{},] }
    ];
    return AkitaNgDevtools;
}());
export { AkitaNgDevtools };
