// @datorama/akita-ng-entity-service/esm5/lib/action-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} storeName
 * @param {?} notifier
 * @return {?}
 */
export function successAction(storeName, notifier) {
    return (/**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var payload = _a.payload, method = _a.method, successMsg = _a.successMsg;
        notifier.dispatch({
            type: 'success',
            storeName: storeName,
            payload: payload,
            method: method,
            successMsg: successMsg
        });
    });
}
/**
 * @param {?} storeName
 * @param {?} notifier
 * @return {?}
 */
export function errorAction(storeName, notifier) {
    return (/**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var payload = _a.payload, method = _a.method, errorMsg = _a.errorMsg;
        notifier.dispatch({
            type: 'error',
            storeName: storeName,
            payload: payload,
            method: method,
            errorMsg: errorMsg
        });
    });
}
