// @datorama/akita-ng-entity-service/esm5/lib/ng-entity-service-notifier.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { Injectable } from '@angular/core';
import { filter } from 'rxjs/operators';
import * as i0 from "@angular/core";
/** @enum {string} */
var HttpMethod = {
    GET: 'GET',
    POST: 'POST',
    PUT: 'PUT',
    PATCH: 'PATCH',
    DELETE: 'DELETE',
};
export { HttpMethod };
/** @type {?} */
export var ofType = (/**
 * @param {?} type
 * @return {?}
 */
function (type) { return filter((/**
 * @param {?} action
 * @return {?}
 */
function (action) { return action.type === type; })); });
/** @type {?} */
export var filterMethod = (/**
 * @param {?} method
 * @return {?}
 */
function (method) {
    return filter((/**
     * @param {?} action
     * @return {?}
     */
    function (action) { return action.method === method; }));
});
/** @type {?} */
export var filterStore = (/**
 * @param {?} name
 * @return {?}
 */
function (name) { return filter((/**
 * @param {?} action
 * @return {?}
 */
function (action) { return action.storeName === name; })); });
var NgEntityServiceNotifier = /** @class */ (function () {
    function NgEntityServiceNotifier() {
        this.dispatcher = new Subject();
        this.action$ = this.dispatcher.asObservable();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NgEntityServiceNotifier.prototype.dispatch = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.dispatcher.next(event);
    };
    NgEntityServiceNotifier.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */ NgEntityServiceNotifier.ngInjectableDef = i0.defineInjectable({ factory: function NgEntityServiceNotifier_Factory() { return new NgEntityServiceNotifier(); }, token: NgEntityServiceNotifier, providedIn: "root" });
    return NgEntityServiceNotifier;
}());
export { NgEntityServiceNotifier };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgEntityServiceNotifier.prototype.dispatcher;
    /** @type {?} */
    NgEntityServiceNotifier.prototype.action$;
}
