// @datorama/akita-ng-entity-service/esm5/lib/ng-entity-service.loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { Injectable } from '@angular/core';
import { filter, map } from 'rxjs/operators';
import { isFunction } from '@datorama/akita';
import { HttpMethod } from './ng-entity-service-notifier';
import * as i0 from "@angular/core";
var NgEntityServiceLoader = /** @class */ (function () {
    function NgEntityServiceLoader() {
        this.dispatcher = new Subject();
        this.loading$ = this.dispatcher.asObservable();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NgEntityServiceLoader.prototype.dispatch = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.dispatcher.next(event);
    };
    /**
     * @param {?=} name
     * @return {?}
     */
    NgEntityServiceLoader.prototype.loadersFor = /**
     * @param {?=} name
     * @return {?}
     */
    function (name) {
        var _this = this;
        /** @type {?} */
        var filterStore = filter((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var storeName = _a.storeName;
            return (name ? storeName === name : true);
        }));
        /** @type {?} */
        var filterMethod = (/**
         * @param {?} mthd
         * @return {?}
         */
        function (mthd) {
            return filter((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var method = _a.method;
                return isFunction(mthd) ? mthd(method) : method === mthd;
            }));
        });
        /** @type {?} */
        var actionBased = (/**
         * @param {?} current
         * @return {?}
         */
        function (current) {
            return _this.loading$.pipe(filterStore, filterMethod(current), map((/**
             * @param {?} action
             * @return {?}
             */
            function (action) { return action.loading; })));
        });
        /** @type {?} */
        var idBased = (/**
         * @param {?} id
         * @param {?} mthd
         * @return {?}
         */
        function (id, mthd) {
            return _this.loading$.pipe(filterStore, filterMethod(mthd), filter((/**
             * @param {?} action
             * @return {?}
             */
            function (action) { return action.entityId === id; })), map((/**
             * @param {?} action
             * @return {?}
             */
            function (action) { return action.loading; })));
        });
        return {
            get$: actionBased(HttpMethod.GET),
            add$: actionBased(HttpMethod.POST),
            update$: actionBased((/**
             * @param {?} method
             * @return {?}
             */
            function (method) { return method === HttpMethod.PUT || method === HttpMethod.PATCH; })),
            delete$: actionBased(HttpMethod.DELETE),
            getEntity: (/**
             * @param {?} id
             * @return {?}
             */
            function (id) { return idBased(id, HttpMethod.GET); }),
            updateEntity: (/**
             * @param {?} id
             * @return {?}
             */
            function (id) { return idBased(id, (/**
             * @param {?} method
             * @return {?}
             */
            function (method) { return method === HttpMethod.PUT || method === HttpMethod.PATCH; })); }),
            deleteEntity: (/**
             * @param {?} id
             * @return {?}
             */
            function (id) { return idBased(id, HttpMethod.DELETE); })
        };
    };
    NgEntityServiceLoader.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */ NgEntityServiceLoader.ngInjectableDef = i0.defineInjectable({ factory: function NgEntityServiceLoader_Factory() { return new NgEntityServiceLoader(); }, token: NgEntityServiceLoader, providedIn: "root" });
    return NgEntityServiceLoader;
}());
export { NgEntityServiceLoader };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgEntityServiceLoader.prototype.dispatcher;
    /** @type {?} */
    NgEntityServiceLoader.prototype.loading$;
}
