// @datorama/akita-ng-entity-service/esm5/lib/ng-entity.service.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EntityService, isDefined } from '@datorama/akita';
import { throwError } from 'rxjs';
import { inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { catchError, finalize, map, tap } from 'rxjs/operators';
import { HttpMethod, NgEntityServiceNotifier } from './ng-entity-service-notifier';
import { NgEntityServiceLoader } from './ng-entity-service.loader';
import { defaultConfig, mergeDeep, NG_ENTITY_SERVICE_CONFIG } from './ng-entity-service.config';
import { isID } from './helpers';
import { errorAction, successAction } from './action-factory';
/** @type {?} */
export var mapResponse = (/**
 * @param {?} config
 * @return {?}
 */
function (config) {
    return map((/**
     * @param {?} res
     * @return {?}
     */
    function (res) { return ((config || {}).mapResponseFn ? config.mapResponseFn(res) : res); }));
});
/**
 * @template S
 */
var /**
 * @template S
 */
NgEntityService = /** @class */ (function (_super) {
    __extends(NgEntityService, _super);
    function NgEntityService(store, config) {
        if (config === void 0) { config = {}; }
        var _this = _super.call(this) || this;
        _this.store = store;
        _this.config = config;
        _this.globalConfig = {};
        _this.http = inject(HttpClient);
        _this.loader = inject(NgEntityServiceLoader);
        _this.notifier = inject(NgEntityServiceNotifier);
        _this.globalConfig = inject(NG_ENTITY_SERVICE_CONFIG);
        _this.mergedConfig = mergeDeep(defaultConfig, _this.globalConfig, config);
        _this.dispatchSuccess = successAction(_this.store.storeName, _this.notifier);
        _this.dispatchError = errorAction(_this.store.storeName, _this.notifier);
        return _this;
    }
    Object.defineProperty(NgEntityService.prototype, "api", {
        get: /**
         * @return {?}
         */
        function () {
            return (this.baseUrl || this.getConfigValue('baseUrl')) + "/" + this.resourceName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgEntityService.prototype, "resourceName", {
        get: /**
         * @return {?}
         */
        function () {
            return this.getConfigValue('resourceName') || this.store.storeName;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} api
     * @return {?}
     */
    NgEntityService.prototype.setBaseUrl = /**
     * @param {?} api
     * @return {?}
     */
    function (api) {
        this.baseUrl = api;
    };
    /**
     * @return {?}
     */
    NgEntityService.prototype.getHttp = /**
     * @return {?}
     */
    function () {
        return this.http;
    };
    /**
     * @return {?}
     */
    NgEntityService.prototype.getConfig = /**
     * @return {?}
     */
    function () {
        return this.mergedConfig;
    };
    /**
     * @template T
     * @param {?=} idOrConfig
     * @param {?=} config
     * @return {?}
     */
    NgEntityService.prototype.get = /**
     * @template T
     * @param {?=} idOrConfig
     * @param {?=} config
     * @return {?}
     */
    function (idOrConfig, config) {
        var _this = this;
        /** @type {?} */
        var url;
        /** @type {?} */
        var isSingle = isID(idOrConfig);
        /** @type {?} */
        var _config = (isSingle ? config : idOrConfig) || {};
        /** @type {?} */
        var method = this.getHttpMethod(HttpMethod.GET);
        if (_config.url) {
            url = _config.url;
        }
        else {
            url = isSingle ? this.api + "/" + idOrConfig : this.api;
        }
        this.loader.dispatch({
            method: method,
            loading: true,
            entityId: isSingle ? idOrConfig : null,
            storeName: this.store.storeName
        });
        return (/** @type {?} */ (this.http[method.toLowerCase()](url, _config).pipe(mapResponse(_config), tap((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (isSingle) {
                _this.store.upsert((/** @type {?} */ (idOrConfig)), data);
            }
            else {
                if (_config.append) {
                    _this.store.add(data);
                }
                else if (_config.upsert) {
                    _this.store.upsertMany(data);
                }
                else {
                    _this.store.set(data);
                }
            }
            _this.dispatchSuccess({
                method: method,
                payload: data,
                successMsg: _config.successMsg
            });
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return _this.handleError(method, error, _config.errorMsg); })), finalize((/**
         * @return {?}
         */
        function () {
            _this.loader.dispatch({
                method: method,
                loading: false,
                storeName: _this.store.storeName
            });
        })))));
    };
    /**
     *
     * Add a new entity - Creates a POST request
     *
     * service.add(entity)
     * service.add(entity, config)
     *
     */
    /**
     *
     * Add a new entity - Creates a POST request
     *
     * service.add(entity)
     * service.add(entity, config)
     *
     * @template T
     * @param {?} entity
     * @param {?=} config
     * @return {?}
     */
    NgEntityService.prototype.add = /**
     *
     * Add a new entity - Creates a POST request
     *
     * service.add(entity)
     * service.add(entity, config)
     *
     * @template T
     * @param {?} entity
     * @param {?=} config
     * @return {?}
     */
    function (entity, config) {
        var _this = this;
        /** @type {?} */
        var method = this.getHttpMethod(HttpMethod.POST);
        this.loader.dispatch({
            method: method,
            loading: true,
            storeName: this.store.storeName
        });
        return (/** @type {?} */ (this.http[method.toLowerCase()](this.resolveUrl(config), entity, config).pipe(mapResponse(config), tap((/**
         * @param {?} entity
         * @return {?}
         */
        function (entity) {
            _this.store.add(entity, config);
            _this.dispatchSuccess({
                method: method,
                payload: entity,
                successMsg: config && config.successMsg
            });
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return _this.handleError(method, error, config && config.errorMsg); })), finalize((/**
         * @return {?}
         */
        function () {
            _this.loader.dispatch({
                method: method,
                loading: false,
                storeName: _this.store.storeName
            });
        })))));
    };
    /**
     *
     * Update an entity - Creates a PUT/PATCH request
     *
     * service.update(id, entity)
     * service.update(id, entity, config)
     *
     */
    /**
     *
     * Update an entity - Creates a PUT/PATCH request
     *
     * service.update(id, entity)
     * service.update(id, entity, config)
     *
     * @template T
     * @param {?} id
     * @param {?} entity
     * @param {?=} config
     * @return {?}
     */
    NgEntityService.prototype.update = /**
     *
     * Update an entity - Creates a PUT/PATCH request
     *
     * service.update(id, entity)
     * service.update(id, entity, config)
     *
     * @template T
     * @param {?} id
     * @param {?} entity
     * @param {?=} config
     * @return {?}
     */
    function (id, entity, config) {
        var _this = this;
        /** @type {?} */
        var method = config && config.method ? config.method : this.getHttpMethod(HttpMethod.PUT);
        this.loader.dispatch({
            method: method,
            loading: true,
            entityId: id,
            storeName: this.store.storeName
        });
        return (/** @type {?} */ (this.http[method.toLocaleLowerCase()](this.resolveUrl(config, id), entity, config).pipe(mapResponse(config), tap((/**
         * @param {?} entity
         * @return {?}
         */
        function (entity) {
            _this.store.update(id, (/** @type {?} */ (entity)));
            _this.dispatchSuccess({
                method: method,
                payload: entity,
                successMsg: config && config.successMsg
            });
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return _this.handleError(method, error, config && config.errorMsg); })), finalize((/**
         * @return {?}
         */
        function () {
            _this.loader.dispatch({
                method: method,
                loading: false,
                entityId: id,
                storeName: _this.store.storeName
            });
        })))));
    };
    /**
     *
     * Delete an entity - Creates a DELETE request
     *
     * service.delete(id)
     * service.delete(id, config)
     *
     */
    /**
     *
     * Delete an entity - Creates a DELETE request
     *
     * service.delete(id)
     * service.delete(id, config)
     *
     * @template T
     * @param {?} id
     * @param {?=} config
     * @return {?}
     */
    NgEntityService.prototype.delete = /**
     *
     * Delete an entity - Creates a DELETE request
     *
     * service.delete(id)
     * service.delete(id, config)
     *
     * @template T
     * @param {?} id
     * @param {?=} config
     * @return {?}
     */
    function (id, config) {
        var _this = this;
        /** @type {?} */
        var method = this.getHttpMethod(HttpMethod.DELETE);
        this.loader.dispatch({
            method: method,
            loading: true,
            entityId: id,
            storeName: this.store.storeName
        });
        return (/** @type {?} */ (this.http[method.toLowerCase()](this.resolveUrl(config, id), config).pipe(mapResponse(config), tap((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            _this.store.remove(id);
            _this.dispatchSuccess({
                method: method,
                payload: res,
                successMsg: config && config.successMsg
            });
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return _this.handleError(method, error, config && config.errorMsg); })), finalize((/**
         * @return {?}
         */
        function () {
            _this.loader.dispatch({
                method: method,
                loading: false,
                entityId: id,
                storeName: _this.store.storeName
            });
        })))));
    };
    /**
     * @private
     * @param {?} type
     * @return {?}
     */
    NgEntityService.prototype.getHttpMethod = /**
     * @private
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return this.mergedConfig.httpMethods[type];
    };
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    NgEntityService.prototype.getConfigValue = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.constructor[key] || this.mergedConfig[key];
    };
    /**
     * @private
     * @param {?} config
     * @param {?=} id
     * @return {?}
     */
    NgEntityService.prototype.resolveUrl = /**
     * @private
     * @param {?} config
     * @param {?=} id
     * @return {?}
     */
    function (config, id) {
        /** @type {?} */
        var customUrl = (config || {}).url;
        if (isDefined(id)) {
            return customUrl || this.api + "/" + id;
        }
        return customUrl || this.api;
    };
    /**
     * @private
     * @param {?} method
     * @param {?} error
     * @param {?} errorMsg
     * @return {?}
     */
    NgEntityService.prototype.handleError = /**
     * @private
     * @param {?} method
     * @param {?} error
     * @param {?} errorMsg
     * @return {?}
     */
    function (method, error, errorMsg) {
        this.dispatchError({
            method: method,
            errorMsg: errorMsg,
            payload: error
        });
        return throwError(error);
    };
    return NgEntityService;
}(EntityService));
/**
 * @template S
 */
export { NgEntityService };
if (false) {
    /** @type {?} */
    NgEntityService.prototype.baseUrl;
    /** @type {?} */
    NgEntityService.prototype.loader;
    /**
     * @type {?}
     * @private
     */
    NgEntityService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    NgEntityService.prototype.notifier;
    /**
     * @type {?}
     * @private
     */
    NgEntityService.prototype.globalConfig;
    /**
     * @type {?}
     * @private
     */
    NgEntityService.prototype.mergedConfig;
    /**
     * @type {?}
     * @private
     */
    NgEntityService.prototype.dispatchSuccess;
    /**
     * @type {?}
     * @private
     */
    NgEntityService.prototype.dispatchError;
    /**
     * @type {?}
     * @protected
     */
    NgEntityService.prototype.store;
    /**
     * @type {?}
     * @private
     */
    NgEntityService.prototype.config;
}
