// @datorama/akita-ng-entity-service/esm2015/lib/ng-entity-service-notifier.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { Injectable } from '@angular/core';
import { filter } from 'rxjs/operators';
import * as i0 from "@angular/core";
/** @enum {string} */
const HttpMethod = {
    GET: 'GET',
    POST: 'POST',
    PUT: 'PUT',
    PATCH: 'PATCH',
    DELETE: 'DELETE',
};
export { HttpMethod };
/** @type {?} */
export const ofType = (/**
 * @param {?} type
 * @return {?}
 */
(type) => filter((/**
 * @param {?} action
 * @return {?}
 */
(action) => action.type === type)));
/** @type {?} */
export const filterMethod = (/**
 * @param {?} method
 * @return {?}
 */
(method) => filter((/**
 * @param {?} action
 * @return {?}
 */
(action) => action.method === method)));
/** @type {?} */
export const filterStore = (/**
 * @param {?} name
 * @return {?}
 */
(name) => filter((/**
 * @param {?} action
 * @return {?}
 */
(action) => action.storeName === name)));
export class NgEntityServiceNotifier {
    constructor() {
        this.dispatcher = new Subject();
        this.action$ = this.dispatcher.asObservable();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    dispatch(event) {
        this.dispatcher.next(event);
    }
}
NgEntityServiceNotifier.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */ NgEntityServiceNotifier.ngInjectableDef = i0.defineInjectable({ factory: function NgEntityServiceNotifier_Factory() { return new NgEntityServiceNotifier(); }, token: NgEntityServiceNotifier, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgEntityServiceNotifier.prototype.dispatcher;
    /** @type {?} */
    NgEntityServiceNotifier.prototype.action$;
}
