// @datorama/akita-ng-entity-service/esm2015/lib/ng-entity-service.loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { Injectable } from '@angular/core';
import { filter, map } from 'rxjs/operators';
import { isFunction } from '@datorama/akita';
import { HttpMethod } from './ng-entity-service-notifier';
import * as i0 from "@angular/core";
export class NgEntityServiceLoader {
    constructor() {
        this.dispatcher = new Subject();
        this.loading$ = this.dispatcher.asObservable();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    dispatch(event) {
        this.dispatcher.next(event);
    }
    /**
     * @param {?=} name
     * @return {?}
     */
    loadersFor(name) {
        /** @type {?} */
        const filterStore = filter((/**
         * @param {?} __0
         * @return {?}
         */
        ({ storeName }) => (name ? storeName === name : true)));
        /** @type {?} */
        const filterMethod = (/**
         * @param {?} mthd
         * @return {?}
         */
        mthd => filter((/**
         * @param {?} __0
         * @return {?}
         */
        ({ method }) => {
            return isFunction(mthd) ? mthd(method) : method === mthd;
        })));
        /** @type {?} */
        const actionBased = (/**
         * @param {?} current
         * @return {?}
         */
        (current) => this.loading$.pipe(filterStore, filterMethod(current), map((/**
         * @param {?} action
         * @return {?}
         */
        action => action.loading))));
        /** @type {?} */
        const idBased = (/**
         * @param {?} id
         * @param {?} mthd
         * @return {?}
         */
        (id, mthd) => this.loading$.pipe(filterStore, filterMethod(mthd), filter((/**
         * @param {?} action
         * @return {?}
         */
        action => action.entityId === id)), map((/**
         * @param {?} action
         * @return {?}
         */
        action => action.loading))));
        return {
            get$: actionBased(HttpMethod.GET),
            add$: actionBased(HttpMethod.POST),
            update$: actionBased((/**
             * @param {?} method
             * @return {?}
             */
            method => method === HttpMethod.PUT || method === HttpMethod.PATCH)),
            delete$: actionBased(HttpMethod.DELETE),
            getEntity: (/**
             * @param {?} id
             * @return {?}
             */
            (id) => idBased(id, HttpMethod.GET)),
            updateEntity: (/**
             * @param {?} id
             * @return {?}
             */
            (id) => idBased(id, (/**
             * @param {?} method
             * @return {?}
             */
            method => method === HttpMethod.PUT || method === HttpMethod.PATCH))),
            deleteEntity: (/**
             * @param {?} id
             * @return {?}
             */
            (id) => idBased(id, HttpMethod.DELETE))
        };
    }
}
NgEntityServiceLoader.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */ NgEntityServiceLoader.ngInjectableDef = i0.defineInjectable({ factory: function NgEntityServiceLoader_Factory() { return new NgEntityServiceLoader(); }, token: NgEntityServiceLoader, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgEntityServiceLoader.prototype.dispatcher;
    /** @type {?} */
    NgEntityServiceLoader.prototype.loading$;
}
