// @datorama/akita/esm2015/src/plugins/stateHistory/entityStateHistoryPlugin.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StateHistoryPlugin } from './stateHistoryPlugin';
import { toBoolean } from '../../toBoolean';
import { skip } from 'rxjs/operators';
import { EntityCollectionPlugin } from '../entityCollectionPlugin';
/**
 * @record
 * @template IDType
 */
export function StateHistoryEntityParams() { }
if (false) {
    /** @type {?|undefined} */
    StateHistoryEntityParams.prototype.entityIds;
}
/**
 * @template State, P
 */
export class EntityStateHistoryPlugin extends EntityCollectionPlugin {
    /**
     * @param {?} query
     * @param {?=} params
     */
    constructor(query, params = {}) {
        super(query, params.entityIds);
        this.query = query;
        this.params = params;
        params.maxAge = toBoolean(params.maxAge) ? params.maxAge : 10;
        this.activate();
        this.selectIds()
            .pipe(skip(1))
            .subscribe((/**
         * @param {?} ids
         * @return {?}
         */
        ids => this.activate(ids)));
    }
    /**
     * @param {?=} ids
     * @return {?}
     */
    redo(ids) {
        this.forEachId(ids, (/**
         * @param {?} e
         * @return {?}
         */
        e => e.redo()));
    }
    /**
     * @param {?=} ids
     * @return {?}
     */
    undo(ids) {
        this.forEachId(ids, (/**
         * @param {?} e
         * @return {?}
         */
        e => e.undo()));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    hasPast(id) {
        if (this.hasEntity(id)) {
            return this.getEntity(id).hasPast;
        }
    }
    /**
     * @param {?} id
     * @return {?}
     */
    hasFuture(id) {
        if (this.hasEntity(id)) {
            return this.getEntity(id).hasFuture;
        }
    }
    /**
     * @param {?} ids
     * @param {?} index
     * @return {?}
     */
    jumpToFuture(ids, index) {
        this.forEachId(ids, (/**
         * @param {?} e
         * @return {?}
         */
        e => e.jumpToFuture(index)));
    }
    /**
     * @param {?} ids
     * @param {?} index
     * @return {?}
     */
    jumpToPast(ids, index) {
        this.forEachId(ids, (/**
         * @param {?} e
         * @return {?}
         */
        e => e.jumpToPast(index)));
    }
    /**
     * @param {?=} ids
     * @return {?}
     */
    clear(ids) {
        this.forEachId(ids, (/**
         * @param {?} e
         * @return {?}
         */
        e => e.clear()));
    }
    /**
     * @param {?=} ids
     * @param {?=} clearHistory
     * @return {?}
     */
    destroy(ids, clearHistory = false) {
        this.forEachId(ids, (/**
         * @param {?} e
         * @return {?}
         */
        e => e.destroy(clearHistory)));
    }
    /**
     * @param {?=} ids
     * @return {?}
     */
    ignoreNext(ids) {
        this.forEachId(ids, (/**
         * @param {?} e
         * @return {?}
         */
        e => e.ignoreNext()));
    }
    /**
     * @protected
     * @param {?} id
     * @return {?}
     */
    instantiatePlugin(id) {
        return (/** @type {?} */ (new StateHistoryPlugin(this.query, this.params, id)));
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    EntityStateHistoryPlugin.prototype.query;
    /**
     * @type {?}
     * @protected
     */
    EntityStateHistoryPlugin.prototype.params;
}
