// @datorama/akita/esm5/src/query.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { queryConfigKey } from './queryConfig';
import { isString } from './isString';
import { isFunction } from './isFunction';
import { isDev } from './env';
import { __queries__ } from './stores';
import { distinctUntilChanged, map } from 'rxjs/operators';
import { compareKeys } from './compareKeys';
/**
 * @template S
 */
var /**
 * @template S
 */
Query = /** @class */ (function () {
    function Query(store) {
        this.store = store;
        this.__store__ = store;
        if (isDev()) {
            // @internal
            __queries__[store.storeName] = this;
        }
    }
    /**
     * @template R
     * @param {?=} project
     * @return {?}
     */
    Query.prototype.select = /**
     * @template R
     * @param {?=} project
     * @return {?}
     */
    function (project) {
        /** @type {?} */
        var mapFn;
        if (isFunction(project)) {
            mapFn = project;
        }
        else if (isString(project)) {
            mapFn = (/**
             * @param {?} state
             * @return {?}
             */
            function (state) { return state[project]; });
        }
        else if (Array.isArray(project)) {
            return (/** @type {?} */ (this.store
                ._select((/**
             * @param {?} state
             * @return {?}
             */
            function (state) { return state; }))
                .pipe(distinctUntilChanged(compareKeys(project)), map((/**
             * @param {?} state
             * @return {?}
             */
            function (state) {
                if (isFunction(project[0])) {
                    return ((/** @type {?} */ (project))).map((/**
                     * @param {?} func
                     * @return {?}
                     */
                    function (func) { return func(state); }));
                }
                return ((/** @type {?} */ (project))).reduce((/**
                 * @param {?} acc
                 * @param {?} k
                 * @return {?}
                 */
                function (acc, k) {
                    acc[(/** @type {?} */ (k))] = state[k];
                    return acc;
                }), {});
            })))));
        }
        else {
            mapFn = (/**
             * @param {?} state
             * @return {?}
             */
            function (state) { return state; });
        }
        return this.store._select(mapFn);
    };
    /**
     * Select the loading state
     *
     * @example
     *
     * this.query.selectLoading().subscribe(isLoading => {})
     */
    /**
     * Select the loading state
     *
     * \@example
     *
     * this.query.selectLoading().subscribe(isLoading => {})
     * @return {?}
     */
    Query.prototype.selectLoading = /**
     * Select the loading state
     *
     * \@example
     *
     * this.query.selectLoading().subscribe(isLoading => {})
     * @return {?}
     */
    function () {
        return this.select((/**
         * @param {?} state
         * @return {?}
         */
        function (state) { return ((/** @type {?} */ (state))).loading; }));
    };
    /**
     * Select the error state
     *
     * @example
     *
     * this.query.selectError().subscribe(error => {})
     */
    /**
     * Select the error state
     *
     * \@example
     *
     * this.query.selectError().subscribe(error => {})
     * @template ErrorType
     * @return {?}
     */
    Query.prototype.selectError = /**
     * Select the error state
     *
     * \@example
     *
     * this.query.selectError().subscribe(error => {})
     * @template ErrorType
     * @return {?}
     */
    function () {
        return this.select((/**
         * @param {?} state
         * @return {?}
         */
        function (state) { return ((/** @type {?} */ (state))).error; }));
    };
    /**
     * Get the store's value
     *
     * @example
     *
     * this.query.getValue()
     *
     */
    /**
     * Get the store's value
     *
     * \@example
     *
     * this.query.getValue()
     *
     * @return {?}
     */
    Query.prototype.getValue = /**
     * Get the store's value
     *
     * \@example
     *
     * this.query.getValue()
     *
     * @return {?}
     */
    function () {
        return this.store._value();
    };
    /**
     * Select the cache state
     *
     * @example
     *
     * this.query.selectHasCache().pipe(
     *   switchMap(hasCache => {
     *     return hasCache ? of() : http().pipe(res => store.set(res))
     *   })
     * )
     */
    /**
     * Select the cache state
     *
     * \@example
     *
     * this.query.selectHasCache().pipe(
     *   switchMap(hasCache => {
     *     return hasCache ? of() : http().pipe(res => store.set(res))
     *   })
     * )
     * @return {?}
     */
    Query.prototype.selectHasCache = /**
     * Select the cache state
     *
     * \@example
     *
     * this.query.selectHasCache().pipe(
     *   switchMap(hasCache => {
     *     return hasCache ? of() : http().pipe(res => store.set(res))
     *   })
     * )
     * @return {?}
     */
    function () {
        return this.store._cache().asObservable();
    };
    /**
     * Whether we've cached data
     *
     * @example
     *
     * this.query.getHasCache()
     *
     */
    /**
     * Whether we've cached data
     *
     * \@example
     *
     * this.query.getHasCache()
     *
     * @return {?}
     */
    Query.prototype.getHasCache = /**
     * Whether we've cached data
     *
     * \@example
     *
     * this.query.getHasCache()
     *
     * @return {?}
     */
    function () {
        return this.store._cache().value;
    };
    Object.defineProperty(Query.prototype, "config", {
        // @internal
        get: 
        // @internal
        /**
         * @return {?}
         */
        function () {
            return this.constructor[queryConfigKey];
        },
        enumerable: true,
        configurable: true
    });
    return Query;
}());
/**
 * @template S
 */
export { Query };
if (false) {
    /** @type {?} */
    Query.prototype.__store__;
    /**
     * @type {?}
     * @protected
     */
    Query.prototype.store;
}
